fs/vfs/fs_eventfd.c

```diff
@@ -0,0 +1,623 @@
+/****************************************************************************
+ * vfs/fs_eventfd.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <stdio.h>
+#include <string.h>
+#include <poll.h>
+#include <errno.h>
+#include <fcntl.h>
+
+#include <debug.h>
+
+#include <sys/ioctl.h>
+#include <sys/eventfd.h>
+
+#include "inode/inode.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define EVENTFD_FMT    "/dev/efd%d"
+
+#ifndef CONFIG_EVENT_FD_NPOLLWAITERS
+/* Maximum number of threads than can be waiting for POLL events */
+#  define CONFIG_EVENT_FD_NPOLLWAITERS 2
+#endif
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* This structure describes the internal state of the driver */
+
+struct eventfd_priv_s
+{
+  sem_t     exclsem;        /* Enforces exclusive access to device counter */
+  sem_t     rdsem;          /* Empty buffer - Reader waits for data write */
+  sem_t     wrsem;          /* Full buffer - Writer waits for data read */
+  uint8_t   crefs;          /* References counts on eventfd (limited to 255) */
+  uint8_t   minor;          /* eventfd minor number */
+  uint8_t   mode_semaphore; /* eventfd mode (semaphore or counter) */
```

Comment:
should use file->f_oflags instead?

Reply:
I'm not sure because I want eventfd mode to be related to the first open (evenfd() syscall). f_oflags can change if someone else opens "/var/event/efdxx".

Reply:
Why we limit the user select the I/O mode? Here is Linux implementation:
https://github.com/torvalds/linux/blob/b29482fde649c72441d5478a4ea2c52c56d97a5e/fs/eventfd.c#L231
You can see it also check file->f_flags.